optax/losses/_self_supervised_test.py

```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
+        q1,
+        z2,
+        q2,
+        z1,
+        symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  def test_single_direction_matches_handmade(self):
+    q = jnp.array(
+        [[0.1, 0.2, 0.3],
+         [0.3, 0.2, 0.1]],
+        dtype=jnp.float32,
+    )
+    z = jnp.array(
+        [[0.4, 0.0, 0.2],
+         [0.1, 0.5, 0.3]],
+        dtype=jnp.float32,
+    )
+
+    def testing_single_direction_byol(q_val, z_val, eps=1e-6):
+      cos = _regression.cosine_similarity(
+          q_val,
+          z_val,
+          epsilon=eps,
+      )
+      loss = 2.0 - 2.0 * cos
+      return jnp.mean(loss)
+
+    handmade_result = testing_single_direction_byol(q, z)
+    result = jax.jit(_self_supervised.byol_loss)(
+        q,
+        z,
+        symmetric=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+
+class SimSiamLossTest(absltest.TestCase):
+
+  def test_symmetric_batched_matches_handmade(self):
+    p1 = jnp.array(
+        [[0.1, 0.2, 0.3],
+         [0.4, 0.5, 0.6]],
+        dtype=jnp.float32,
+    )
+    p2 = jnp.array(
+        [[0.2, 0.1, 0.0],
+         [0.3, 0.2, 0.1]],
+        dtype=jnp.float32,
+    )
+    z1 = jnp.array(
+        [[0.5, 0.4, 0.3],
+         [0.2, 0.1, 0.0]],
+        dtype=jnp.float32,
+    )
+    z2 = jnp.array(
+        [[0.3, 0.2, 0.1],
+         [0.6, 0.5, 0.4]],
+        dtype=jnp.float32,
+    )
+
+    def testing_simsiam_loss(p1_val, z2_val, p2_val, z1_val, eps=1e-6):
+      cos_12 = _regression.cosine_similarity(
+          p1_val,
+          z2_val,
+          epsilon=eps,
+      )
+      cos_21 = _regression.cosine_similarity(
+          p2_val,
+          z1_val,
+          epsilon=eps,
+      )
+      loss_12 = -cos_12
+      loss_21 = -cos_21
+      loss = 0.5 * (loss_12 + loss_21)
+      return jnp.mean(loss)
+
+    handmade_result = testing_simsiam_loss(p1, z2, p2, z1)
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p1,
+        z2,
+        p2,
+        z1,
+        symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  def test_single_direction_matches_handmade(self):
+    p = jnp.array(
+        [[0.1, 0.2, 0.3],
+         [0.3, 0.2, 0.1]],
+        dtype=jnp.float32,
+    )
+    z = jnp.array(
+        [[0.4, 0.0, 0.2],
+         [0.1, 0.5, 0.3]],
+        dtype=jnp.float32,
+    )
+
+    def testing_single_direction_simsiam(p_val, z_val, eps=1e-6):
+      cos = _regression.cosine_similarity(
+          p_val,
+          z_val,
+          epsilon=eps,
+      )
+      loss = -cos
+      return jnp.mean(loss)
+
+    handmade_result = testing_single_direction_simsiam(p, z)
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p,
+        z,
+        symmetric=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+
+class DinoLossTest(absltest.TestCase):
+
+  def test_single_view_matches_handmade(self):
```

Comment:
rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)